File: Buddy-iOS-SDK/Test/BuddySDKTests/IdentityUnitTests.m
/*
 * Copyright (C) 2012 Buddy Platform, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

#import "IdentityUnitTests.h"
#import "TestBuddySDK.h"
#import "BuddyCallbackParams.h"
#import "BuddyDataResponses.h"
#import "BuddyBoolResponse.h"
#import "BuddyClient.h"
#import "BuddyIdentityItemSearchResult.h"


@implementation IdentityUnitTests

static NSString *AppName = @"Buddy iOS SDK test app";
static NSString *AppPassword = @"8C9E044D-7DB7-42DE-A376-16460B58008E";
static bool bwaiting = false;
static NSString *Token = @"UT-76444f9f-4a4b-4d3d-ba5c-7a82b5dbb5a5";

@synthesize buddyClient;
@synthesize user;

- (void)setUp
{
    [super setUp];
    
    self.buddyClient = [[BuddyClient alloc] initClient:AppName
                                           appPassword:AppPassword
                                            appVersion:@"1"
                                  autoRecordDeviceInfo:TRUE];
    
    STAssertNotNil(self.buddyClient, @"IdentityUnitTests failed buddyClient nil");
}

- (void)tearDown
{
    [super tearDown];
    
    self.buddyClient = nil;
    self.user = nil;
}

- (void)alogin
{
    [self.buddyClient login:Token  callback:[^(BuddyAuthenticatedUserResponse *response)
    {
        if (response.isCompleted)
        {
            self.user = response.result;
            NSLog(@"alogin OK user: %@", self.user.toString);
        }
        else
        {
            STFail(@"alogin failed !response.isCompleted");
        }
        bwaiting = false;
    } copy]];
}

- (void)waitloop
{
    NSDate *loopTil = [NSDate dateWithTimeIntervalSinceNow:15];
    
    while (bwaiting && [loopTil timeIntervalSinceNow] > 0)
        [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:loopTil];
}

- (void)testIdentityParsing
{
    NSArray *resArray = [TestBuddySDK GetTextFileData:@"Test_Identity"];
    
    bwaiting = true;
    [self alogin];
    [self waitloop];
    
    if (!self.user)
    {
        STFail(@"testIdentityParsing login failed.");
        return;
    }
    
    NSArray *dataout = [self.user.identityValues performSelector:@selector(makeIdentityList:) withObject:resArray];
    if ([dataout count] != 3)
    {
        STFail(@"testIdentityParsing failed makeIdentityList expected 3");
    }
    
    NSArray *dataout1 = [self.user.identityValues performSelector:@selector(makeIdentitySearchList:) withObject:resArray];
    if ([dataout1 count] != 3)
    {
        STFail(@"testIdentityParsingNoData makeIdentitySearchList expects 3 ");
    }
}

- (void)testIdentityParsingMyList
{
    NSArray *resArray = [TestBuddySDK GetTextFileData:@"Test_IdentityMyList"];
    
    bwaiting = true;
    [self alogin];
    [self waitloop];
    
    if (!self.user)
    {
        STFail(@"testIdentityParsingMyList login failed.");
        return;
    }
    
    NSArray *dataout = [self.user.identityValues performSelector:@selector(makeIdentityList:) withObject:resArray];
    
    if ([dataout count] != 3)
    {
        STFail(@"testIdentityParsingMyList failed makeIdentityList expected 3 ");
    }
    
    NSArray *dataout1 = [self.user.identityValues performSelector:@selector(makeIdentitySearchList:) withObject:resArray];
    if ([dataout1 count] != 3)
    {
        STFail(@"testIdentityParsingMyList makeIdentitySearchList expects 3 ");
    }
    
    if ([dataout1 count] > 0)
    {
        BuddyIdentityItemSearchResult *bSrc = [dataout1 objectAtIndex:0];
        BOOL found = bSrc.found;
        NSNumber *num = bSrc.belongsToUserId;
        NSString *val = bSrc.value;
        NSDate *date = bSrc.createdOn;
        date = nil;
        num = nil;
        val = nil;
        found = FALSE;
    }
}

- (void)testIdentityParsingNoData
{
    NSArray *resArray = [TestBuddySDK GetTextFileData:@"Test_NoData"];
    
    bwaiting = true;
    [self alogin];
    [self waitloop];
    
    if (!self.user)
    {
        STFail(@"testIdentityParsingNoData login failed.");
        return;
    }
    
    NSArray *dataout = [self.user.identityValues performSelector:@selector(makeIdentityList:) withObject:resArray];
    if ([dataout count] != 0)
    {
        STFail(@"testIdentityParsingNoData failed makeIdentityList expected 0");
    }
    
    NSArray *dataout1 = [self.user.identityValues performSelector:@selector(makeIdentitySearchList:) withObject:resArray];
    if ([dataout1 count] != 0)
    {
        STFail(@"testIdentityParsingNoData makeIdentitySearchList expects 0 ");
    }
}

- (void)testIdentityParsingBadData
{
    NSArray *resArray = [TestBuddySDK GetTextFileData:@"Test_IdentityBad"];
    
    bwaiting = true;
    [self alogin];
    [self waitloop];
    
    if (!self.user)
    {
        STFail(@"testIdentityParsingBadData login failed.");
        return;
    }
    
    NSArray *dataout = [self.user.identityValues performSelector:@selector(makeIdentityList:) withObject:resArray];
    if ([dataout count] != 3)
    {
        STFail(@"testIdentityParsingBadData failed makeIdentityList expected 3");
    }
    
    NSArray *dataout1 = [self.user.identityValues performSelector:@selector(makeIdentitySearchList:) withObject:resArray];
    if ([dataout1 count] != 3)
    {
        STFail(@"testIdentityParsingBadData makeIdentitySearchList expects 3");
    }
}

- (void)testIdentityParsingEmptyData
{
    NSArray *resArray = [TestBuddySDK GetTextFileData:@"Test_EmptyData"];
    
    bwaiting = true;
    [self alogin];
    [self waitloop];
    
    if (!self.user)
    {
        STFail(@"testIdentityParsingEmptyData login failed.");
        return;
    }
    
    NSArray *dataout = [self.user.identityValues performSelector:@selector(makeIdentityList:) withObject:resArray];
    if ([dataout count] != 0)
    {
        STFail(@"testIdentityParsingEmptyData expects 0");
    }
    
    NSArray *dataout1 = [self.user.identityValues performSelector:@selector(makeIdentitySearchList:) withObject:resArray];
    if ([dataout1 count] != 0)
    {
        STFail(@"testIdentityParsingEmptyData makeIdentitySearchList expects 0");
    }
}

- (void)atestIdentityRemove:(BuddyAuthenticatedUser *)buddyuser
{
    [self IdentityRemoveTest:buddyuser];
}

- (void)IdentityRemoveTest:(BuddyAuthenticatedUser *)buddyuser
{
    [buddyuser.identityValues remove:@"John Smith" 
                            callback:[^(BuddyBoolResponse *response)
                            {
                                if (response.isCompleted && response.result == TRUE)
                                {
                                    NSLog(@"IdentityRemoveTest OK");
                                }
                                else
                                {
                                    STFail(@"IdentityRemoveTest failed !response.isCompleted || !response.result");
                                }
                                bwaiting = false;
                            } copy]];
}


- (void)atestStateIdentityRemove:(BuddyAuthenticatedUser *)buddyuser
{
    [self StateIdentityRemoveTest:buddyuser ];
}

- (void)StateIdentityRemoveTest:(BuddyAuthenticatedUser *)buddyuser 
{
    [buddyuser.identityValues remove:@"John Smith" 
                            callback:[^(BuddyBoolResponse *response)
                            {
                                NSString *_state = (NSString *) response.stringResult;
                                if ([_state isEqualToString:@"John Smith"])
                                {
                                    NSLog(@"StateIdentityRemoveTest OK %@", _state);
                                }
                                else
                                {
                                    STFail(@"StateIdentityRemoveTest failed !response.state == John Smith");
                                }
                                bwaiting = false;
                            } copy]];
}


@end

File: Buddy-Android-SDK/Tests/Tests/src/com/buddy/sdk/unittests/IdentityUnitTests.java
/* Copyright (C) 2012 Buddy Platform, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package com.buddy.sdk.unittests;

import java.util.List;
import com.buddy.sdk.Identity;
import com.buddy.sdk.IdentityItem;
import com.buddy.sdk.Callbacks.OnCallback;
import com.buddy.sdk.IdentityItemSearchResult;
import com.buddy.sdk.responses.Response;

import com.buddy.sdk.responses.ListResponse;
import com.buddy.sdk.web.BuddyHttpClientFactory;

public class IdentityUnitTests extends BaseUnitTest {
    @Override
    protected void setUp() throws Exception {
        super.setUp();

        createAuthenticatedUser();
    }

    public void testGetIdentityValues() {
        String jsonResponse = readDataFromFile("DataResponses/IdentityUnitTests-GetAll.json");
        BuddyHttpClientFactory.addDummyResponse(jsonResponse);

        Identity identity = testAuthUser.getIdentityValues();

        identity.getAll(null, new OnCallback<ListResponse<IdentityItem>>() {

            @Override
            public void OnResponse(ListResponse<IdentityItem> response, Object state) {
                assertNotNull(response);

                List<IdentityItem> list = response.getList();
                assertNotNull(list);

                IdentityItem item = list.get(0);
                assertEquals("Test Value", item.getValue());
            }

        });

    }

    public void testSearchIdentityValue() {
        String jsonResponse = readDataFromFile("DataResponses/IdentityUnitTests-CheckValues.json");
        BuddyHttpClientFactory.addDummyResponse(jsonResponse);

        Identity identity = testAuthUser.getIdentityValues();

        identity.checkForValues("Test Value", null,
                new OnCallback<ListResponse<IdentityItemSearchResult>>() {

                    @Override
                    public void OnResponse(ListResponse<IdentityItemSearchResult> response,
                            Object state) {
                        assertNotNull(response);

                        List<IdentityItemSearchResult> list = response.getList();
                        assertNotNull(list);

                        IdentityItemSearchResult item = list.get(0);
                        assertEquals("Test Value", item.getValue());
                    }

                });
    }

    public void testAddIdentityValue() {
        String jsonResponse = readDataFromFile("DataResponses/GenericSuccessResponse.json");
        BuddyHttpClientFactory.addDummyResponse(jsonResponse);

        Identity identity = testAuthUser.getIdentityValues();

        identity.add("Test Value", null, new OnCallback<Response<Boolean>>() {

            @Override
            public void OnResponse(Response<Boolean> response, Object state) {
                assertNotNull(response);
                assertTrue(response.getResult());
            }

        });
    }

    public void testRemoveIdentityValue() {
        String jsonResponse = readDataFromFile("DataResponses/GenericSuccessResponse.json");
        BuddyHttpClientFactory.addDummyResponse(jsonResponse);

        Identity identity = testAuthUser.getIdentityValues();

        identity.remove("Test Value", null, new OnCallback<Response<Boolean>>() {

            @Override
            public void OnResponse(Response<Boolean> response, Object state) {
                assertNotNull(response);
                assertTrue(response.getResult());
            }

        });
    }
}
